### PR TITLE
NHibernate unique constraints migration guide 

### PR DESCRIPTION
**Pull request by tmasternak**
Connects to https://github.com/Particular/NServiceBus.NHibernate/issues/280.

ping @MarcinHoppe @SzymonPobiega

### CODE REVIEW

**Comment by MarcinHoppe** [comment]
@tmasternak These SQL statements are SQL Server specific, right (e.g. `nonclustered`)? Is there a way to make them work both on SQL Server and Oracle?

**Comment by tmasternak** [comment]
@MarcinHoppe good point. I forgot about Oracle 😒. I would probably need to setup an instance on my box to validate that.

**Comment by MarcinHoppe** [comment]
@tmasternak I can do this for you in a spare moment but some things are done different (e.g. executing dynamic SQL). Can you create separate sections for Oracle and I can fill the snippets later?

**Comment by tmasternak** [comment]
@MarcinHoppe done

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ * For all saga mapping table
+    * Check if it holds any duplicated rows
+    * If any exist manually merge the duplicates
+    * Add unique constraint on correlation property column
+ * Update endpoint to latest patch release
+ * Deploy the new version
+
+### Checking if table hold duplicates on Microsoft SQL Server
+
+Duplicates in saga mapping table can be detected using following query. It requires providing `sagaMappingTableName` and `correlationPropetyColumnName`:
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: Better as:

> The following query detects duplicate rows in the saga data table. It requires `sagaDataTableName` and `correlationPropertyColumnName`:

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+```sql
+declare @sagaMappingTableName nvarchar(max) = ...
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: sagaMappingTableName  -> sagaDataTableName

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+declare @correlationPropretyColumnName nvarchar(max) = ...
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: correlationPropretyColumnName  -> correlationPropertyColumnName

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+declare @sql nvarchar(max)
+
+select @sql = 'select ' + @correlationPropertyColumnName + ', count(*) as SagaRows from ' + @sagaMappingTableName + ' group by ' + @correlationPropertyColumnName + ' having count(*) > 1'
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: Use `SagaRows` alias in the `HAVING` clause.

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+exec sp_executeSQL @sql
+
+```
+
+If the query returns any results each row will represent single logical saga instance. First column of the result will show the correlation property value for logical saga instance.
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: This

> If the query returns any results each row will represent single logical saga instance.

would read better as:

> Each row in the result set represents a single logical saga instance that was duplicated.

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+### Checking if table hold duplicates on Oracle
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: I'd rather see Oracle and SQL Server subsections under `Checking for duplicate correlation property values` section. This way only SQL snippets would be different.

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+
+### Add unique constraint on correlation property column on Microsoft SQL Server
+
+NOTE: If adding unique constraint fails with `The CREATE UNIQUE INDEX statement terminated because a duplicate key was found for the object name ...` message please make sure that all duplicated rows detected have been merged.
+
+
+Unique constraint on correlation property column can be added with following query. It requires providing `sagaMappingTableName` and `correlationPropetyColumnName`:
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: Better as:

> The following statement adds a unique constraint on the correlation property column.

Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: sagaMappingTableName -> sagaDataTableName

Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: correlationPropetyColumnName -> correlationPropertyColumnName

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+```sql
+declare @sagaMappingTableName nvarchar(max) = ...
```
Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: sagaMappingTableName -> sagaDataTableName

File: nservicebus/upgrades/nhibernate-saga-index-fix.md (hunk 2 above)
Comment (SimonCropp) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: i would prefer this in a snippet. so we get a better editing experience in VS

**Comment by SimonCropp** [comment]
should this be in the menu somewhere?

File: nservicebus/upgrades/nhibernate-saga-index-fix.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
+---
+title: NHibernate Persistence - unique constraints on saga mapping tables
```
Comment (tmasternak) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: There is constraint on the length of article title (70 chars). Currently it's 65 :P so the suggested version will be to long.

Comment (MarcinHoppe) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: Then maybe: 

> NHibernate Persistence - Missing unique constraints

?

Comment (tmasternak) [inline nservicebus/upgrades/nhibernate-saga-index-fix.md]: I like it

**Comment by MarcinHoppe** [comment]
This needs a rebase.

File: Snippets/NHibernate/NHibernate_7/UpgradeGuides/MissingUniqueConstraintOnSagaDataTables/unique_constraints.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
+﻿
+
+startcode DetectSagaDataDuplicates_MsSqlServer
```
Comment (SimonCropp) [inline Snippets/NHibernate/NHibernate_7/UpgradeGuides/MissingUniqueConstraintOnSagaDataTables/unique_constraints.sql]: should this should be a sql comment. so it is still a valid sql statement

**Pull request merged**